design-documents/frontend/technical-vision.md

```diff
@@ -0,0 +1,140 @@
+
+We should prepare to achieve as much code sharing as possible, regardless of environment. JavaScript is the only language that can run natively in all browsers; yet, it also [transcends the browser](https://cdb.reacttraining.com/universal-javascript-4761051b7ae9), running on servers and [even native mobile applications](https://facebook.github.io/react-native/). As such, the golden path for sharing code starts with JavaScript. By writing components in a decoupled way, we don't limit ourselves as to where those components can run.
+
+On top of sharing code between different environments, we should also be concerned about sharing code between projects. For this, we should introduce a business logic repository that includes common helper functions – logic that we're likely to use between projects.
+
+### Business Logic Repository
+1. Written and [documented](https://github.com/Microsoft/tsdoc#tsdoc) in [TypeScript][].
+1. [Semantic versioning](https://semver.org/) to ensure that progress can be made without breaking things for existing - consumers.
+1. 100% unit test coverage via [Jest][] for **consistency**.
+
+For **performance** reasons, JavaScript should only be used for cheap operations on the front end, reserving expensive operations for a services layer. It's worth calling out that we could achieve even more code sharing if this API were written in JavaScript. For practical reasons, however, it makes more sense to leverage internal PHP resources for this task.
+
+## Browser Support
+All users should have a **consistent** user experience. For the best **performance**, those with the latest browsers should not be penalized by incurring an additional payload hit than those with older browsers. To achieve this goal, we download just the [polyfills](https://remysharp.com/2010/10/08/what-is-a-polyfill) needed for our sites, tailored to each browser via [Polyfill.io](https://polyfill.io) feature detection.
+
+> We test browser support with a test suite for each feature. A browser is considered compliant only if it passes all the tests, so this is a reasonably comprehensive test rather than a naïve feature-detect.
+
+_TODO: Investigate building two separate [TypeScript][] builds, one that compiles down to ES5 and another lighter-weight modern build, specifically for modern browsers. Depending on how well Polyfill.io works, the modern build might be all we need._
+
+## React UI Component Library
+
+_Consistency, Productivity, Scalability, Maintainability, Accessibility._
+
+We should lean against downloading various UI components from different sources for the following reasons:
+
+- **Inconsistent** implementations, documentation, theming and customization.
+- Limited **accessibility** and support.
+
+[KendoReact | UI for React Developers](https://www.telerik.com/kendo-react-ui/) is a great option for those wishing to get React UI components from the same source. Still, there will be times when you need to download a 3rd-party package.

```

Comment:
I think we ended up discussing reasons offline that Kendo wasn't going to work out well for us and our OSS code. I think we should remove the advocacy for it here since this document will be used more often than not for OSS projects.

Reply:
Right, not for us, but this document is public, so I had to remove certain information, remember? That said, KendoReact is still a great alternative for those who have no other options. The ultimate point here is that if they have the ability to pay for the product, it's much better to get all your components from a single source than to have completely different paradigms scattered everywhere.

Reply:
> The ultimate point here is that if they have the ability to pay for the product, it's much better to get all of your components from a single source than to have completely different paradigms scattered everywhere.

I think you just wrote up the better generic description to replace a recommendation of a specific product :)

Reply:
If you're going to pay for a React component library though, I really do recommend KendoReact. It's a solid product.

Reply:
Addressed in 4f625a47b50e0c0ac67ccbf5fed03eff17b94d40